File: android/domain/src/main/kotlin/com/controlldeck/domain/Enums.kt
package com.controlldeck.domain

/** Platform a device runs on. Mirrors protocol/PROTOCOL.md HELLO.platform. */
enum class Platform {
    ANDROID,
    WINDOWS,
}

/**
 * Capability enum mirrors protocol/PROTOCOL.md §3.4. Closed set for the MVP;
 * unknown wire values are tolerated at the protocol layer and never reach
 * this enum (they're dropped before mapping into the domain).
 */
enum class Capability {
    BRIGHTNESS,
    VOLUME,
    MUTE,
    MEDIA_PLAY_PAUSE,
    MEDIA_NEXT,
    MEDIA_PREVIOUS,
    APP_LAUNCH,
}

/** Widget kinds per docs/ARCHITECTURE.md §3. */
enum class WidgetType {
    SLIDER_BRIGHTNESS,
    SLIDER_VOLUME,
    BUTTON_MUTE,
    BUTTON_MEDIA_PLAY_PAUSE,
    BUTTON_MEDIA_NEXT,
    BUTTON_MEDIA_PREVIOUS,
    APP_LAUNCH,
}

/** The capability a widget type requires on its target device. */
fun WidgetType.requiredCapability(): Capability = when (this) {
    WidgetType.SLIDER_BRIGHTNESS -> Capability.BRIGHTNESS
    WidgetType.SLIDER_VOLUME -> Capability.VOLUME
    WidgetType.BUTTON_MUTE -> Capability.MUTE
    WidgetType.BUTTON_MEDIA_PLAY_PAUSE -> Capability.MEDIA_PLAY_PAUSE
    WidgetType.BUTTON_MEDIA_NEXT -> Capability.MEDIA_NEXT
    WidgetType.BUTTON_MEDIA_PREVIOUS -> Capability.MEDIA_PREVIOUS
    WidgetType.APP_LAUNCH -> Capability.APP_LAUNCH
}

/** Group semantics kind. See docs/ARCHITECTURE.md §4. */
enum class GroupKind {
    RELATIVE_SLIDER,
    ABSOLUTE_TOGGLE,
    ABSOLUTE_MEDIA,
}

/** Reconnect behavior for a group member. See docs/ARCHITECTURE.md §5. */
enum class ReconnectPolicy {
    SYNC_GROUP_STATE,
    KEEP_DEVICE_STATE,
    NO_ACTION,
}

/** Device connection state as tracked by the local State Manager. */
enum class ConnectionState {
    ONLINE,
    OFFLINE,
}

/** Media playback state. */
enum class MediaPlaybackState {
    PLAYING,
    PAUSED,
}

File: windows/src/ControlDeck.Domain/Enums.cs
namespace ControlDeck.Domain;

/// <summary>Device platform, per protocol/PROTOCOL.md §3.1 ("ANDROID" | "WINDOWS").</summary>
public enum Platform
{
    Android,
    Windows
}

/// <summary>
/// Closed capability enum for the MVP, per protocol/PROTOCOL.md §3.4. Receivers
/// must tolerate unknown future values elsewhere on the wire (handled in the
/// Protocol layer); this enum only needs to represent the values this codebase
/// understands.
/// </summary>
public enum Capability
{
    Brightness,
    Volume,
    Mute,
    MediaPlayPause,
    MediaNext,
    MediaPrevious,
    AppLaunch
}

/// <summary>Widget kinds, per docs/ARCHITECTURE.md §3.</summary>
public enum WidgetType
{
    SliderBrightness,
    SliderVolume,
    ButtonMute,
    ButtonMediaPlayPause,
    ButtonMediaNext,
    ButtonMediaPrevious,
    AppLaunch
}

/// <summary>Group semantics, per docs/ARCHITECTURE.md §4.</summary>
public enum GroupKind
{
    RelativeSlider,
    AbsoluteToggle,
    AbsoluteMedia
}

/// <summary>Reconnect behavior for a group member, per docs/ARCHITECTURE.md §5.</summary>
public enum ReconnectPolicy
{
    SyncGroupState,
    KeepDeviceState,
    NoAction
}

/// <summary>Per-device online/offline state, per docs/ARCHITECTURE.md §3/§7.</summary>
public enum ConnectionState
{
    Online,
    Offline
}

/// <summary>Media playback state, per protocol/PROTOCOL.md §3.5.</summary>
public enum MediaState
{
    Playing,
    Paused
}
